<a href="https://colab.research.google.com/github/nathlia/Aulas2023/blob/RECUPERA%C3%87%C3%83O-DE-INFORMA%C3%87%C3%83O-E-PERSONALIZA%C3%87%C3%83O-NA-WEB-(DPADP0152)/Entrega1_nathalia.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Nathalia F. Fernandes

### Link relatorio: https://github.com/nathlia/Aulas2023/blob/RECUPERA%C3%87%C3%83O-DE-INFORMA%C3%87%C3%83O-E-PERSONALIZA%C3%87%C3%83O-NA-WEB-(DPADP0152)/Entrega1_nathalia.ipynb

# Imports e iniciando as files

In [12]:
import numpy as np
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity

In [13]:
from google.colab import files

# Use the files.upload() method to upload files
uploaded = files.upload()


importando os stopwords

In [14]:
## Library imports
import numpy as np
import pandas as pd

from sklearn.feature_extraction.text import CountVectorizer

import os, glob, re, sys, random, unicodedata, collections
from tqdm import tqdm
from functools import reduce
import nltk
nltk.download('stopwords')
from collections import Counter

from nltk.corpus import stopwords
from nltk.stem import RSLPStemmer
from nltk.tokenize import sent_tokenize , word_tokenize

STOP_WORDS = set(stopwords.words('english'))

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [15]:
stop_words_list = list(STOP_WORDS)
print(stop_words_list)


['he', 'these', 'didn', 'then', "doesn't", 'o', 'into', 'shouldn', 'doesn', 'herself', 'those', 'if', 'out', 'there', 'before', 'few', 'd', 'while', 'such', "wouldn't", 'at', 'during', 'does', 'here', "weren't", 'are', 'has', "mightn't", 've', 'had', 'over', 'what', 's', 'ma', 'only', "didn't", 'hadn', 'until', 'me', 'ourselves', 'you', "you've", 'that', 'for', "shan't", 'she', 'between', 'mustn', 'shan', 'won', 'not', 'which', 't', 'theirs', 'as', 'is', "you'd", 'be', 'where', 'too', 'couldn', 'because', "it's", 'haven', 'can', 'aren', "needn't", 'who', 'why', 'y', 'hers', 'both', 'himself', 'been', 'on', 'we', 'your', 'did', 'further', "mustn't", "hasn't", 'have', 'should', "she's", 'again', 'hasn', 'each', 'myself', 'of', 'against', 'through', 'and', 'no', 'wasn', 'wouldn', 'all', 'our', 'isn', "haven't", "you'll", 'do', 'about', 'yourself', 'now', 'than', 'i', 'a', 'its', 'whom', 'up', "hadn't", 'or', 'them', 'itself', 'from', 'nor', "you're", 'having', "should've", 'ours', 'their'

In [18]:
from google.colab import drive
drive.mount('/content/drive/')

Mounted at /content/drive/


mounted google drive para acessar os textos, escolhi a pasta tech e importei para o drive

In [20]:
import os

# Path to your Google Drive directory
google_drive_path = '/content/drive/MyDrive/ENTREGA1/tech'

# List the files and subdirectories in your Google Drive directory
file_list = os.listdir(google_drive_path)

# Count the number of files
file_count = len(file_list)

# Print the count
print(f"Number of files in your Google Drive directory: {file_count}")

Number of files in your Google Drive directory: 401


In [22]:
import random


In [27]:
# select and read 10 random files
sample_texts = file_list
# sample_texts = random.sample(file_list, 100)

docs = []
for fname in sample_texts:
    with open(google_drive_path + '/' + fname , "r", encoding="ISO-8859-1") as file:
        text = file.read()
    docs.append(text)

print("Finished reading files in:", google_drive_path)

Finished reading files in: /content/drive/MyDrive/ENTREGA1/tech


In [28]:
# count term frequency using CountVectorizer from scikit-learn
## limiting number of words just for illustrating the concept

vec = CountVectorizer(max_features=20, stop_words=stop_words_list)
X = vec.fit_transform(docs)
df = pd.DataFrame(X.toarray(), columns=vec.get_feature_names_out())
files_names = [file.split('/')[-1] for file in sample_texts]
df['file'] = files_names
df = df.set_index('file')

print(df)

         also  could  digital  games  many  mobile  mr  music  net  new  one  \
file                                                                           
374.txt     1      1        2      3     0       0   0      2    0    3    3   
164.txt     1      2        2      0     3       0   0      0    2    0    2   
293.txt     0      1        0      0     2       0   2      1    0    4    4   
095.txt     4      2        1     15     4       9   0      1    0    1    0   
359.txt     6      1        1      3     0       2   0      2    0    0    0   
...       ...    ...      ...    ...   ...     ...  ..    ...  ...  ...  ...   
347.txt     2      5        5      0     1       3   2      5    1    0    3   
020.txt     2      0        0      0     1       0   1      0    0    6    1   
010.txt     2      2        1      0     0       0   1      0    2    1    0   
006.txt     7      0        0      0     0       0   3      0    0    1    1   
005.txt     1      0        1      2    

# Consultas

Elabore 6 exemplos de consulta.

In [29]:
# Define queries
queries = [
    "Document security and information leakage in Microsoft Word",
    "Consumer Electronics Show (CES) 2005 highlights",
    "Digital games and technology",
    "Impact of technology on people",
    "Web links and virus",
    "Cyber-crime",
]

for q in queries:
    print(q)

Document security and information leakage in Microsoft Word
Consumer Electronics Show (CES) 2005 highlights
Digital games and technology
Impact of technology on people
Web links and virus
Cyber-crime


In [30]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity

# Utils

Calcula Similiariedade do Coseno

In [33]:
def calcule_cosine_similarities(query_matrix, document_matrix):
    cosine = cosine_similarity(query_matrix, document_matrix)
    return cosine

Print resultado da busca

Considera o documento como valido dependendo do valor da similiariedade de coseno, ordena em ordem decrescente o valor de similiaredade dos documentos validos.  

In [34]:
# Print matching document titles and file names based on similarity threshold
def print_similarity(queries, cosine_similarities, documents, file_names, threshold):
    # Calcular a média das similaridades para cada consulta
    similaridade_media = [sum(sim_scores) / len(sim_scores) for sim_scores in cosine_similarities]

    highest_similarity = -1.0  # Initialize with a value lower than any possible similarity
    lowest_similarity = 2.0  # Initialize with a value higher than any possible similarity
    highest_similarity_info = None
    lowest_similarity_info = None

    for i, query in enumerate(queries):
        print('---' * 30)
        print(f"⭐ Query {i + 1}: {query} ⭐")
        print(f"Similaridade Média: {similaridade_media[i]:.4f}")  # Adicione esta linha
        print('---' * 30)
        similar_documents = []

        for j, sim_score in enumerate(cosine_similarities[i]):
            if sim_score > threshold:  # Adjust the similarity threshold as needed
                document_title = documents[j].split('\n')[0]  # Extract the first line (title)
                similar_documents.append((document_title, file_names[j], sim_score))

                # Update highest and lowest similarity scores and their corresponding info
                if sim_score > highest_similarity:
                    highest_similarity = sim_score
                    highest_similarity_info = (document_title, file_names[j], sim_score)
                if sim_score < lowest_similarity:
                    lowest_similarity = sim_score
                    lowest_similarity_info = (document_title, file_names[j], sim_score)

        if similar_documents:
            num = 0
            similar_documents.sort(key=lambda x: x[2], reverse=True)
            for title, file_name, score in similar_documents:
                num = num + 1;
                print(f"📄 {num} File: {file_name} ∽ Similarity: {score:.4f} ✒️Title: {title}")
        else:
            print("No matching titles found.")
        print("\n")

    print(f"\nHighest Similarity:")
    if highest_similarity_info:
        title, file_name, score = highest_similarity_info
        print(f"Highest Similarity Title: {title}")
        print(f"Highest Similarity File Name: {file_name}")
        print(f"Highest Similarity Score: {score:.4f}")

    print(f"\nLowest Similarity:")
    if lowest_similarity_info:
        title, file_name, score = lowest_similarity_info
        print(f"Lowest Similarity Title: {title}")
        print(f"Lowest Similarity File Name: {file_name}")
        print(f"Lowest Similarity Score: {score:.4f}")

# Primeira rodada: sem retirar stopword e sem "stemming"


Cria um objeto TfidfVectorizer chamado de **vectorizer** que converte os dados do text em um vetor **TF-IDF** (Term Frequency-Inverse Document Frequency), que é uma **representação numérica de documentos de texto**.

A lista de documentos é transformada em uma matriz usando o metodo de fit_trasform do vectorizer.

Calcula similiariedade de cosenos entre os vetores das queries e os vetores dos documentos.

In [35]:
def find_matching_titles(queries, documents, file_names):
    # Create a TfidfVectorizer
    vectorizer = TfidfVectorizer()

    # Fit and transform the documents to obtain TF-IDF vectors
    tfidf_matrix = vectorizer.fit_transform(documents)

    # Transform the queries into TF-IDF vectors using the same vectorizer
    query_tfidf_matrix = vectorizer.transform(queries)

    # Calculate cosine similarities between queries and documents
    cosine_similarities = calcule_cosine_similarities(query_tfidf_matrix, tfidf_matrix)

    print_similarity(queries, cosine_similarities, documents, file_names, 0)

# Example usage
find_matching_titles(queries, docs, file_list)

------------------------------------------------------------------------------------------
⭐ Query 1: Document security and information leakage in Microsoft Word ⭐
Similaridade Média: 0.0401
------------------------------------------------------------------------------------------
📄 1 File: 086.txt ∽ Similarity: 0.4431 ✒️Title: Warning over Windows Word files
📄 2 File: 308.txt ∽ Similarity: 0.1926 ✒️Title: Microsoft makes anti-piracy move
📄 3 File: 083.txt ∽ Similarity: 0.1926 ✒️Title: Microsoft makes anti-piracy move
📄 4 File: 003.txt ∽ Similarity: 0.1833 ✒️Title: Microsoft seeking spyware trojan
📄 5 File: 036.txt ∽ Similarity: 0.1833 ✒️Title: Microsoft seeking spyware trojan
📄 6 File: 369.txt ∽ Similarity: 0.1766 ✒️Title: Microsoft plans 'safer ID' system
📄 7 File: 060.txt ∽ Similarity: 0.1643 ✒️Title: Microsoft releases patches
📄 8 File: 007.txt ∽ Similarity: 0.1639 ✒️Title: Microsoft releases bumper patches
📄 9 File: 020.txt ∽ Similarity: 0.1587 ✒️Title: Security scares spark brows

## Comentário:

Possui numero variaveis de similaridade entre os documentos, o que afirma que os documentos possuem um nivel diferente de relevancia entre as buscas.

A maior similaridade encontrada foi na Busca 1 "Document security and information leakage in Microsoft Word", sendo ela 0.4431 do documento "Warning over Windows Word files", que foi feita especificamente para esse texto.

A unicas buscas que não mostraram todos os documentos foram:
*   Busca 2: "Consumer Electronics Show (CES) 2005 highlights"
*   Busca 6: "Cyber-crime"

Que foram retiradas diretamente do texto.

Coloquei uma condicao de só mostrar documentos com similaridade maior que 0. Mesmo assim 4 das 6 buscam possuem todos os documentos como similares.

Comparando com o resultado do codigo exemplo do ChatGPT, nota-se que esse método possui resultados de similiriaedade maiores que o metodo do ChatGPT, mas continua mostrando a mesma quantidade de documentos.


### Exemplo ChatGPT da aula

In [ ]:
# Initialize the TfidfVectorizer
vectorizer = TfidfVectorizer()

# Fit and transform your documents and queries
tfidf_matrix = vectorizer.fit_transform(docs + queries)

# Calculate the cosine similarities between the queries and documents
cosine_similarities = cosine_similarity(tfidf_matrix[-len(queries):], tfidf_matrix[:-len(queries)])


# Print only the titles of matching documents for each query
for i, query in enumerate(queries):
    sorted_indices = cosine_similarities[i].argsort()[::-1]
    print('---' * 20)
    print(f"☆ Query {i + 1}: {query} ☆")
    print('---' * 20)
    print("\nMatching Document Titles:")
    for idx in sorted_indices:
        document_title = docs[idx].split('\n')[0]  # Extract the first line (title)
        if cosine_similarities[i][idx] >  0:
            print(f"Similarity: {cosine_similarities[i][idx]:.4f} - Title: {document_title}")
    print('\n')

#  Segunda rodada: retirando stopwords e sem "stemming"




In [37]:
def find_matching_titles_segunda(queries, documents, file_names):
    # Create a TfidfVectorizer with stopwords removed
    vectorizer = TfidfVectorizer(stop_words=stop_words_list)

    # Fit and transform the documents to obtain TF-IDF vectors
    tfidf_matrix = vectorizer.fit_transform(documents)

    # Transform the queries into TF-IDF vectors using the same vectorizer
    query_tfidf_matrix = vectorizer.transform(queries)

    # Calculate cosine similarities between queries and documents
    cosine_similarities = calcule_cosine_similarities(query_tfidf_matrix, tfidf_matrix)

    print_similarity(queries, cosine_similarities, documents, file_names, 0)

# Example usage
find_matching_titles_segunda(queries, docs, file_list)

------------------------------------------------------------------------------------------
⭐ Query 1: Document security and information leakage in Microsoft Word ⭐
Similaridade Média: 0.0192
------------------------------------------------------------------------------------------
📄 1 File: 086.txt ∽ Similarity: 0.4693 ✒️Title: Warning over Windows Word files
📄 2 File: 308.txt ∽ Similarity: 0.1919 ✒️Title: Microsoft makes anti-piracy move
📄 3 File: 083.txt ∽ Similarity: 0.1919 ✒️Title: Microsoft makes anti-piracy move
📄 4 File: 003.txt ∽ Similarity: 0.1716 ✒️Title: Microsoft seeking spyware trojan
📄 5 File: 036.txt ∽ Similarity: 0.1716 ✒️Title: Microsoft seeking spyware trojan
📄 6 File: 369.txt ∽ Similarity: 0.1685 ✒️Title: Microsoft plans 'safer ID' system
📄 7 File: 060.txt ∽ Similarity: 0.1601 ✒️Title: Microsoft releases patches
📄 8 File: 007.txt ∽ Similarity: 0.1597 ✒️Title: Microsoft releases bumper patches
📄 9 File: 020.txt ∽ Similarity: 0.1540 ✒️Title: Security scares spark brows

### Comentário

Com a remoção das stopwords, nota-se que nenhuma das buscas retornam todos os documentos, e todas estão com um número total diferente de documentos. A quantidade de resultados diminuiu consideravelmente.

O valor da similaridade dos primeiros itens aumentam levemente também, as palavras-chaves possuem um impacto maior. Mas no caso da busca 4 "Impact of technology on people", ela caiu. A queda após a remoção das stopwords pode indicar que essa consulta pode ser menos específica ou que as palavras-chave  não estão tão bem representadas nos documentos indexados.

Maior similaridade continua a mesma, enquando a menor mudou. O que faz sentido pois a maior é determinada pelo documento que melhor corresponde a consulta, e a menor pode ser mais afetada com a remoção das stopwords.


# Terceira rodada: retirando stopwords e  com "stemming"

In [38]:
import nltk
from nltk.stem import SnowballStemmer

# Inicialize o stemmer para a língua desejada (por exemplo, inglês)
stemmer = SnowballStemmer("english")

# Aplicar stemming às consultas e documentos
stemmed_queries = [" ".join([stemmer.stem(word) for word in query.split()]) for query in queries]


In [39]:
def find_matching_titles_terceira(queries, documents, file_names):

    stemmed_docs = [" ".join([stemmer.stem(word) for word in doc.split()]) for doc in documents]

    # Create a TfidfVectorizer with stopwords removed
    vectorizer = TfidfVectorizer(stop_words=stop_words_list)

    # Fit and transform the documents to obtain TF-IDF vectors
    tfidf_matrix = vectorizer.fit_transform(stemmed_docs)

    # Transform the stemmed queries into TF-IDF vectors using the same vectorizer
    query_tfidf_matrix = vectorizer.transform(stemmed_queries)

    # Calculate cosine similarities between stemmed queries and documents
    cosine_similarities = calcule_cosine_similarities(query_tfidf_matrix, tfidf_matrix)

    # Print matching document titles and file names based on similarity threshold
    print_similarity(queries, cosine_similarities, documents, file_names, 0.1)


find_matching_titles_terceira(queries, docs, file_list)

------------------------------------------------------------------------------------------
⭐ Query 1: Document security and information leakage in Microsoft Word ⭐
Similaridade Média: 0.0216
------------------------------------------------------------------------------------------
📄 1 File: 086.txt ∽ Similarity: 0.6069 ✒️Title: Warning over Windows Word files
📄 2 File: 151.txt ∽ Similarity: 0.2058 ✒️Title: 'Blog' picked as word of the year
📄 3 File: 308.txt ∽ Similarity: 0.2019 ✒️Title: Microsoft makes anti-piracy move
📄 4 File: 083.txt ∽ Similarity: 0.2019 ✒️Title: Microsoft makes anti-piracy move
📄 5 File: 003.txt ∽ Similarity: 0.1853 ✒️Title: Microsoft seeking spyware trojan
📄 6 File: 036.txt ∽ Similarity: 0.1853 ✒️Title: Microsoft seeking spyware trojan
📄 7 File: 369.txt ∽ Similarity: 0.1783 ✒️Title: Microsoft plans 'safer ID' system
📄 8 File: 060.txt ∽ Similarity: 0.1612 ✒️Title: Microsoft releases patches
📄 9 File: 007.txt ∽ Similarity: 0.1608 ✒️Title: Microsoft releases bumper p

### Comentário

Após remover as stopwords e aplicar o steamming, os resultados ficaram consideravelmente menores.

A similaridade dos dos documentos de maior ranking aumentaram levemente na maioria das consultas, isso indica que a remoção das stopwords e aplicação do steamming tiveram um impacto positivo na classificação.

As pontuações médias para cada consulta também  mudaram, em alguns casos ficou maior, em outros menores.

Documentos de maior similariedade continua o mesmo, enquanto o menor mudou novamente, mas é o maior numero na categoria de Lowest Similarity Score.

No geral, aumentou a qualidade de pesquisa.

# Comparação

### **Primeira Rodada (Sem retirar stopwords e sem stemming):**

A similaridade entre documentos variou, indicando diferentes níveis de relevância.
Quatro das seis buscas retornaram todos os documentos como similares, sugerindo baixa especificidade.
Comparado ao método de exemplo do ChatGPT, sua abordagem gerou maiores valores de similaridade, mas manteve a mesma quantidade de documentos.


### **Segunda Rodada (Retirando stopwords e sem stemming):**

A remoção de stopwords reduziu significativamente a quantidade de resultados em todas as consultas.
Alguns documentos de maior ranking tiveram um leve aumento na similaridade, indicando um foco maior nas palavras-chave.
No entanto, em uma consulta específica, a similaridade caiu, sugerindo que essa consulta pode ser menos específica.
A maior similaridade ainda permaneceu a mesma, enquanto a menor variou, o que era esperado após a remoção das stopwords.

### **Terceira Rodada (Retirando stopwords e com stemming):**

Após a aplicação de stemming, os resultados foram consideravelmente menores.
A maioria das consultas teve um leve aumento na similaridade dos documentos de maior ranking, indicando um impacto positivo na classificação.
As pontuações médias das consultas também mudaram, algumas aumentaram e outras diminuíram.
A maior similaridade permaneceu a mesma, mas a menor mudou novamente.

A segunda e terceira rodadas parecem ter produzido resultados mais específicos e relevantes, especialmente após a aplicação de stemming. Isso sugere que essas técnicas podem melhorar a eficácia das buscas ao focar nas palavras-chave e reduzir a ambiguidade das consultas.

### Similariedade Média

| **Round**          | **Highest Similarity**                   | **Lowest Similarity**                    |
|--------------------|-----------------------------------------|-----------------------------------------|
| Primeira rodada    | Title: Warning over Windows Word files | Title: Hi-tech posters guide commuters  |
|                    | File: 086.txt                           | File: 230.txt                           |
|                    | Score: 0.4431                           | Score: 0.0017                           |
| Segunda rodada     | Title: Warning over Windows Word files | Title: Mac Mini heralds mini revolution |
|                    | File: 086.txt                           | File: 336.txt                           |
|                    | Score: 0.4693                           | Score: 0.0027                           |
| Terceira rodada    | Title: Warning over Windows Word files | Title: Gritty return for Prince of Persia|
|                    | File: 086.txt                           | File: 218.txt                           |
|                    | Score: 0.6069                           | Score: 0.1004                           |


 A remoção de stopwords e a aplicação de stemming ajudaram a aumentar a similaridade media das consultas, o que é um sinal positivo de que as consultas estão gerando resultados mais relevantes e precisos.

# Conceitos

### Bag of Words (BoW)

The Bag of Words approach is like making a list of all the unique words you find in these books, without caring about the order or how many times each word appears. It's like counting how many times each word shows up but not worrying about the order in which they appear.

In [41]:
BoW = vectorizer.fit_transform(file_list)

print(BoW)

  (0, 401)	0.15668736734446556
  (0, 373)	0.9876482516132251
  (1, 163)	0.9876482516132251
  (1, 401)	0.15668736734446556
  (2, 292)	0.9876482516132251
  (2, 401)	0.15668736734446556
  (3, 94)	0.9876482516132251
  (3, 401)	0.15668736734446556
  (4, 358)	0.9876482516132251
  (4, 401)	0.15668736734446556
  (5, 387)	0.9876482516132251
  (5, 401)	0.15668736734446556
  (6, 335)	0.9876482516132251
  (6, 401)	0.15668736734446556
  (7, 1)	0.9876482516132251
  (7, 401)	0.15668736734446556
  (8, 22)	0.9876482516132251
  (8, 401)	0.15668736734446556
  (9, 154)	0.9876482516132251
  (9, 401)	0.15668736734446556
  (10, 304)	0.9876482516132251
  (10, 401)	0.15668736734446556
  (11, 11)	0.9876482516132251
  (11, 401)	0.15668736734446556
  (12, 376)	0.9876482516132251
  :	:
  (388, 401)	0.15668736734446556
  (389, 28)	0.9876482516132251
  (389, 401)	0.15668736734446556
  (390, 325)	0.9876482516132251
  (390, 401)	0.15668736734446556
  (391, 75)	0.9876482516132251
  (391, 401)	0.15668736734446556
  (392

### N-grams

This modification will create a matrix where each column represents a bigram (a combination of two words).

In [42]:
n_gramsVectorizer = CountVectorizer(ngram_range=(2, 2))

n_grams = n_gramsVectorizer.fit_transform(file_list)

print(n_grams)

  (0, 373)	1
  (1, 163)	1
  (2, 292)	1
  (3, 94)	1
  (4, 358)	1
  (5, 387)	1
  (6, 335)	1
  (7, 1)	1
  (8, 22)	1
  (9, 154)	1
  (10, 304)	1
  (11, 11)	1
  (12, 376)	1
  (13, 232)	1
  (14, 386)	1
  (15, 356)	1
  (16, 42)	1
  (17, 173)	1
  (18, 111)	1
  (19, 309)	1
  (20, 182)	1
  (21, 99)	1
  (22, 84)	1
  (23, 316)	1
  (24, 37)	1
  :	:
  (376, 319)	1
  (377, 64)	1
  (378, 148)	1
  (379, 322)	1
  (380, 399)	1
  (381, 398)	1
  (382, 3)	1
  (383, 274)	1
  (384, 342)	1
  (385, 227)	1
  (386, 214)	1
  (387, 39)	1
  (388, 383)	1
  (389, 28)	1
  (390, 325)	1
  (391, 75)	1
  (392, 207)	1
  (393, 77)	1
  (394, 49)	1
  (395, 7)	1
  (396, 346)	1
  (397, 19)	1
  (398, 9)	1
  (399, 5)	1
  (400, 4)	1


### HashingVectorizer
The n_features parameter determines the number of features in the output matrix. HashingVectorizer doesn't build a vocabulary like CountVectorizer.

In [43]:
from sklearn.feature_extraction.text import HashingVectorizer

hashingV = HashingVectorizer(n_features=1000)

hashing = hashingV.transform(file_list)

print(hashing)

  (0, 371)	-0.7071067811865475
  (0, 869)	-0.7071067811865475
  (1, 110)	0.7071067811865475
  (1, 869)	-0.7071067811865475
  (2, 845)	-0.7071067811865475
  (2, 869)	-0.7071067811865475
  (3, 568)	0.7071067811865475
  (3, 869)	-0.7071067811865475
  (4, 487)	0.7071067811865475
  (4, 869)	-0.7071067811865475
  (5, 50)	-0.7071067811865475
  (5, 869)	-0.7071067811865475
  (6, 285)	0.7071067811865475
  (6, 869)	-0.7071067811865475
  (7, 821)	0.7071067811865475
  (7, 869)	-0.7071067811865475
  (8, 551)	-0.7071067811865475
  (8, 869)	-0.7071067811865475
  (9, 449)	0.7071067811865475
  (9, 869)	-0.7071067811865475
  (10, 2)	-0.7071067811865475
  (10, 869)	-0.7071067811865475
  (11, 12)	-0.7071067811865475
  (11, 869)	-0.7071067811865475
  (12, 581)	-0.7071067811865475
  :	:
  (388, 975)	-0.7071067811865475
  (389, 631)	0.7071067811865475
  (389, 869)	-0.7071067811865475
  (390, 629)	0.7071067811865475
  (390, 869)	-0.7071067811865475
  (391, 701)	0.7071067811865475
  (391, 869)	-0.7071067811865

### Word vs. Character N-grams

You can specify whether you want word or character N-grams using the analyzer parameter. For example, to work with character bigrams (N=2) with word boundaries (like " j" for "jump"), you can do this:

This will create a matrix where each column represents a character bigram with word boundaries.*italicized text*

In [44]:
bigrams_with_boundariesV = CountVectorizer(analyzer='char_wb', ngram_range=(2, 2))

bigrams = bigrams_with_boundariesV.fit_transform(file_list)

print(bigrams)

  (0, 3)	1
  (0, 47)	1
  (0, 88)	1
  (0, 50)	1
  (0, 5)	1
  (0, 117)	1
  (0, 118)	1
  (0, 116)	1
  (1, 50)	1
  (1, 5)	1
  (1, 117)	1
  (1, 118)	1
  (1, 116)	1
  (1, 1)	1
  (1, 24)	1
  (1, 77)	1
  (2, 5)	1
  (2, 117)	1
  (2, 118)	1
  (2, 116)	1
  (2, 2)	1
  (2, 38)	1
  (2, 109)	1
  (2, 39)	1
  (3, 5)	1
  :	:
  (397, 29)	1
  (398, 5)	1
  (398, 117)	1
  (398, 118)	1
  (398, 116)	1
  (398, 0)	1
  (398, 8)	1
  (398, 18)	1
  (398, 6)	1
  (399, 5)	1
  (399, 117)	1
  (399, 118)	1
  (399, 116)	1
  (399, 0)	1
  (399, 72)	1
  (399, 7)	1
  (399, 13)	1
  (400, 5)	1
  (400, 117)	1
  (400, 118)	1
  (400, 116)	1
  (400, 0)	1
  (400, 61)	1
  (400, 7)	1
  (400, 12)	1
